### PR TITLE
Feat/add token

### PR DESCRIPTION
**Pull request by Fawadkhan**
Token feature

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -5,45 +5,80 @@
...
-  await login(client, username, password);
-  info('Logged in to WAPM.');
+// login with username and password 
+const loginWithUsername = async (username: string, password: string) => {
 
-  //Publish to WAPM
-  await publish(directory);
-  info(`Published the package located in ${directory}.`);
+  
+    const directory = getInput('directory') != '' ? getInput('directory') : process.cwd();
```
Comment (syrusakbary) [inline src/index.ts]: We should only call `getInput` once.

Such as:

```suggestion
    const directory = getInput('directory') || process.cwd();
```

Comment (syrusakbary) [reply src/index.ts]: (we need to test that my suggestion works)

File: src/wapm.ts [hunk 2]
```diff
@@ -88,6 +88,37 @@ export const login = async (client: Client, username: string, password: string)
...
+  //Authenticate to WAPM
+  // const {data} = await client.mutate({
+  //   mutation: LoginTokenDocument,
+  //   variables: {
+  //     userToken
+  //   }
+  // });
+
+  //Extract the data
+  // const token = data!.tokenAuth!.refreshToken!;
```
Comment (syrusakbary) [inline src/wapm.ts]: can we delete this?

File: src/wapm.ts [hunk 3]
```diff
@@ -88,6 +88,37 @@ export const login = async (client: Client, username: string, password: string)
...
+  const exitCode = await exec('wapm', ['config', 'set', 'registry.token', userToken]);
+
+  //Ensure the child was successful
+  if (exitCode != 0)
+  {
+    //Fail the step
+    setFailed(`Failed to set WAPM token, WAPM exited with ${exitCode}!`);
+
+    //Crash
+    process.exit(1);
```
Comment (syrusakbary) [inline src/wapm.ts]: We shouldn't exit here (imagine you are testing this function case, it would be impossible to test it with unit testing in tests).
We should probably throw an error and catch it on the `main` one, exiting in `main` with `1`

File: src/wapm.spec.ts [hunk 4]
```diff
@@ -6,12 +6,14 @@
 import test from 'ava';
 import {existsSync} from 'fs';
 import {getClient} from './apollo';
-import {getLocation, getRegistryUrl, login, publish, setRegistryUrl} from './wapm';
+import {getLocation, getRegistryUrl, login, publish, setRegistryUrl, tokenLogin} from './wapm';
 import {loadEnv, loggedIn, logout, resolveFixture} from './test-utilities';
-
+// WAPM_REGISTRY_TOKEN=wap_bcc06873de6ee1cbab3e1ca9144eef93923b74bb5a8c8c72189acb47e00a3622
```
Comment (syrusakbary) [inline src/wapm.spec.ts]: Remove me

File: src/index.ts [hunk 5]
```diff
@@ -5,45 +5,80 @@
...
+    }
+    const client = getClient(registry);
+  
+    //Login to WAPM
+    await login(client, username, password);
+    info('Logged in to WAPM.');
+  
+    //Publish to WAPM
+    await publish(directory);
+    info(`Published the package located in ${directory}.`);
```
Comment (syrusakbary) [inline src/index.ts]: This seems the wrong place to publish, why do we need to publish when calling the login function?
We should move this call to the `main` function

File: src/index.ts [hunk 6]
```diff
@@ -5,45 +5,80 @@
...
+    }
+    const client = getClient(registry);
+  
+    //Login to WAPM
+    await tokenLogin(client, userToken);
+    info('Logged in to WAPM.');
+  
+    //Publish to WAPM
+    await publish(directory);
+    info(`Published the package located in ${directory}.`);
```
Comment (syrusakbary) [inline src/index.ts]: This seems the wrong place to publish, why do we need to publish when calling the login function?
We should move this call to the `main` function

**Comment by Wakeful-Cloud** [comment]
@syrusakbary Let me know if you need any help with this

**Comment by syrusakbary** [comment]
Hey @Wakeful-Cloud if you could also chime in and review this PR prior to merging that would be great!

We are trying to add support for token-based authentication into `wapm publish`

File: README.md [hunk 7]
```diff
@@ -28,17 +28,22 @@
   with:
     username: ${{ secrets.WAPM_USERNAME }}
     password: ${{ secrets.WAPM_PASSWORD }}
+    WAPM_REGISTRY_TOKEN: ${{ secrets.WAPM_REGISTRY_TOKEN }}
```
Comment (Wakeful-Cloud) [inline README.md]: This needs to be added to `action.yml` and the key should probably be `token` to match the existing pattern. I'm not sure if the secret should be `WAPM_TOKEN` (Which again follows the existing pattern) or `WAPM_REGISTRY_TOKEN` (To distinguish that the token is specifically for the registry). @syrusakbary may have more insight than I do.

File: README.md [hunk 8]
```diff
@@ -28,17 +28,22 @@
...
 ```
 
 ## Input
 
 | Name        | Required | Description                                                   |
-| ----------- | -------- | ------------------------------------------------------------- |
+| ----------- | -------- | ------------------------------------------------------------  |
 | `username`  | ✅       | WAPM account username                                         |
 | `password`  | ✅       | WAPM account password                                         |
+|               OR       |
```
Comment (Wakeful-Cloud) [inline README.md]: This is a little ambiguous in the rendered markdown - it looks like the input is either `username` and `password` or `token`, `directory`, and `registry` are accepted. I would change it to
```markdown
| Name        | Required | Description                                                    |
| ----------- | -------- | -------------------------------------------------------------- |
| `username`  | 〰️       | WAPM account username (Instead of the `token`)                |
| `password`  | 〰️       | WAPM account password (Instead of the `token`)                |
| `token`     | 〰️       | WAPM account token (Instead of the `username`/`password`)     |
| `directory` | ❌       | Package directory (Defaults to the current working directory) |
| `registry`  | ❌       | The registry url (Defaults to `https://registry.wapm.io/`)    |
```

File: src/index.ts [hunk 9]
```diff
@@ -5,45 +5,80 @@
...
   //Publish to WAPM
   await publish(directory);
   info(`Published the package located in ${directory}.`);
-
+ 
   //Exit
   process.exit();
 };
 
+
```
Comment (Wakeful-Cloud) [inline src/index.ts:32]: I don't think these should be separate functions considering how similar they are (and we aren't unit testing them). I would refactor this so the `login` and `tokenLogin` calls remain in the `main` function.

File: src/index.ts [hunk 10]
```diff
@@ -5,45 +5,80 @@
...
 
 const main = async () =>
 {
   //Get input
-  const username = getInput('username', {
-    required: true
-  });
+  const username = getInput('username');
+  const password = getInput('password');
+  const userToken = getInput('token');
```
Comment (Wakeful-Cloud) [inline src/index.ts:15]: In the event someone specifies a `username`, `password`, and `token`, the `username` and `password` could potentially be leaked (since `setSecret` would never be called for either the `username` or `password`, only for `token`). I don't see any harm in always calling `setSecret` for the `username`, `password`, and `token`, regardless of whether or not the user passed them and it would cover many edge cases.

File: src/test-utilities.ts [hunk 11]
```diff
@@ -19,7 +19,7 @@ export const loadEnv = (name: string) =>
   //If null, crash
   if (value == null || value.length == 0)
   {
-    throw new Error(`Environment variable ${name} not provided!`);
+    return ''
```
Comment (Wakeful-Cloud) [inline src/test-utilities.ts:23]: I know throwing an error breaks tests, but I'm hesitant about just returning an empty string. Perhaps a compromise is to log a warning and then return an empty string? That way it's still obvious if someone forgets to specify credentials when they run unit tests.

File: src/wapm.spec.ts [hunk 12]
```diff
@@ -55,42 +56,87 @@ test.serial('registry url', async ctx =>
   });
 });
 
-test('will login', async ctx =>
+if(userToken) {
```
Comment (Wakeful-Cloud) [inline src/wapm.spec.ts:59]: I'm not sure we should conditionally run unit tests like this, since that would leave certain code untested depending on what credentials the tester provided (I would remove the `if` statement to always run all unit tests). However, that will require someone to generate a token and add it to this repository as a secret (I would use the same WAPM account already used for testing). @syrusakbary When we migrated this repository from the cloud-cnc organization, did you update the `WAPM_DEV_USERNAME` and `WAPM_DEV_PASSWORD` secrets? If not, I can generate the token.

File: src/wapm.spec.ts [hunk 13]
```diff
@@ -6,12 +6,13 @@
...
 import {existsSync} from 'fs';
 import {getClient} from './apollo';
-import {getLocation, getRegistryUrl, login, publish, setRegistryUrl} from './wapm';
+import {getLocation, getRegistryUrl, login, publish, setRegistryUrl, tokenLogin} from './wapm';
 import {loadEnv, loggedIn, logout, resolveFixture} from './test-utilities';
-
 //Environment variables
 const username = loadEnv('WAPM_USERNAME');
 const password = loadEnv('WAPM_PASSWORD');
+const userToken = loadEnv('WAPM_REGISTRY_TOKEN');
```
Comment (Wakeful-Cloud) [inline src/wapm.spec.ts]: If the token secret name in `.github/workflows/test.yml` is updated from `WAPM_REGISTRY_TOKEN`, make sure to update this too.

File: src/index.ts [hunk 14]
```diff
@@ -5,45 +5,80 @@
...
+  if(userToken) {
+    try {
+      loginWithToken(userToken)
```
Comment (Wakeful-Cloud) [inline src/index.ts]: I would remove the try-catch block here, and call `setFailed` and `process.exit(1)` in the catch block at the bottom of this file instead. That way, we handle errors with the same code, regardless if the error originated from `login`, `loginWithUsername`, or `publish`.

File: src/wapm.spec.ts [hunk 15]
```diff
@@ -55,42 +56,87 @@ test.serial('registry url', async ctx =>
...
+  test('will publish with token', async ctx =>
   {
-    //Publish (Dry run)
-    await publish(fixture, true);
+    //Login
+    const client = getClient(registry);
+    await tokenLogin(client, userToken);
+  
+    //Get the fixture
+    const fixture = resolveFixture();
```
Comment (Wakeful-Cloud) [inline src/wapm.spec.ts:88]: Perhaps combine all `resolveFixture` calls by calling it before individual test cases?

File: src/wapm.spec.ts [hunk 16]
```diff
@@ -55,42 +56,87 @@ test.serial('registry url', async ctx =>
...
+  
+    //Assert that it will not throw an error
+    await ctx.notThrowsAsync(async () =>
+    {
+      //Publish (Dry run)
+      await publish(fixture, true);
+    });
+  
+    //Logout
+    // logout();
```
Comment (Wakeful-Cloud) [inline src/wapm.spec.ts:98]: I know I didn't originally do this, but I would call `logout` before and after every test to prevent tests from interfering with each other (You could just call `logout` in the [`test.beforeEach`](https://github.com/wasmerio/wapm-publish/blob/9a642e9d48b040505d250f6bfb1cce67bf77e59e/src/wapm.spec.ts#L23) and [`test.afterEach`](https://github.com/wasmerio/wapm-publish/blob/9a642e9d48b040505d250f6bfb1cce67bf77e59e/src/wapm.spec.ts#L27) hooks).

**Review by Wakeful-Cloud — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (syrusakbary) [reply README.md]: This should be:

```suggestion
    token: ${{ secrets.WAPM_REGISTRY_TOKEN }}
```

File: src/index.ts (hunk 10 above)
Comment (syrusakbary) [reply src/index.ts:15]: Agree. We need to address this

File: src/wapm.spec.ts (hunk 12 above)
Comment (syrusakbary) [reply src/wapm.spec.ts:59]: Agree, we need to remove the if to always run the unit tests.
I believe the WAPM_DEV_USERNAME and WAPM_DEV_PASSWORD were also updated.

**Comment by syrusakbary** [comment]
This is now fixed/done in `master`. Closing the PR
